Add mapping generator for macOS

mohd-akram opened the pull request:
Fixes #1701

kimgr commented:
Very nice! Python is so modern it makes my head spin :-)

I'll need some time to review this in detail, but overall looks great.

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+            public_path = p.with_name(p.name[1:])
+            if public_path in paths:
+                p = public_path
+            else:
```
I'm not sure I follow the loop here --

* if `p` starts with `_`, we strip off the first `_`
* if the stripped path exists in the public paths, we update `p` to the stripped name and goto 1
* if it doesn't, we break out of the loop

Ah, so for `___wctype.h`, we'll keep stripping off `_` until we're at `wctype.h`.

But for `_static_assert.h` there is no public header `static_assert.h`, so that will be left alone?

It seems a little sensitive if there's a case like `___header.h` (three underscores), `_header.h` (one underscore) and `header.h` (none, public header), but I suppose that doesn't exist.

I wonder if this would be easier to express as:
```
public_path = p.parent.with_name(p.name.lstrip('_'))
if public_path in paths:
    p = public_path
```
 or does that do something else?

kimgr replied:
Ah, right, the subsequent rules might do something. Alright, makes sense.

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+        # arm -> machine
+        if p.parent.name in {'arm', 'arm64', 'i386'}:
+            public_path = p.parent.with_name('machine') / p.name
+            if public_path in paths:
```
But if not, should that be an error/warning?

kimgr replied:
(These warning suggestions can probably be dropped if they're very noisy by default. But it sounded like these were both things that should never happen?)

mohd-akram replied:
It should only be a warning if all the attempts fail since a subsequent step might manage to map the header.

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+            if public_path in paths:
+                p = public_path
+
+        # sys/time.h -> time.h
```
This comment looks like it no longer applies.

mohd-akram replied:
It still applies, but that's just one example. I've clarified that this is just an example.

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+                break
+
+        # we don't know this one
+        if p.name.startswith('_'):
```
Again, maybe a little warning might be nice, so users notice when new headers are overlooked? There's a recent example in the libstdcxx generator, that prints to stderr

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+
+    includes.sort(key=itemgetter(0))
+
+    print(f'# Generated using {Path(__file__).name}')
```
Might as well use entire sys.argv here so we can see which source directory for reproducibility. Fancy example in libstdcxx mapgen, which does shell line wrapping, probably not necessary here.

mohd-akram replied:
Only reason I didn't do that is because the include path on macOS...is a bit long (`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include`) :D

kimgr replied:
Ouch! Well, I guess it's worth overflowing the 80-col limit for traceability 🙂

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+
+    includes = []
+
+    mappings = {
```
Can make this a global constant, eg PATTERN_MAPPINGS or something

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,123 @@
+
+    paths = set(root.glob('**/*.h'))
+
+    includes = []
```
I feel like this would be better named 'mappings'

mohd-akram commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,129 @@
+
+
+PATTERN_MAPPINGS = {
+    'pthread_*': 'pthread.h',
+    'sys/_types/_fd_*': 'sys/select.h',
+    'sys/_types/_in_*': 'netinet/in.h',
+    'sys/_types/_o_*': 'fcntl.h',
+    'sys/_types/_s_*': 'sys/stat.h',
+    'sys/_types/_seek_*': 'stdio.h',
```
We can get rid of all of these special cases by adding more symbols to iwyu_include_picker.cc, since they're all POSIX symbols with standard locations. The benefit of that is that it would also work on other systems. See #1706.

kimgr replied:
I wonder if it makes sense to move ahead with these independently, and merge this first?

I'd like to see some test output from #1706, and that's currently blocked by #1700.

mohd-akram replied:
Sure, sounds good to me.

kimgr replied:
I'll try to get to it tomorrow.

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+
+##===--- iwyu-mapgen-macos.py ---------------------------------------------===##
```
Sorry for thinking about naming so late, but I think this should be `iwyu-mapgen-apple-libc.py` -- the Apple Libc seems to be used across macOS, iOS, tvOS and watchOS.

mohd-akram replied:
You're right, that definitely is more accurate.

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,129 @@
+#
+##===----------------------------------------------------------------------===##
+
+""" Generate mappings for macOS libc.
```
macOS libc -> Apple Libc

kimgr replied:
And let's update commit message title too

kimgr commented on mapgen/iwyu-mapgen-macos.py:
```diff
@@ -0,0 +1,129 @@
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description='Generate IWYU mappings for macOS libc.'
```
Apple Libc

kimgr commented:
Thank you!

firewave commented:
I integrated it into the Cppcheck CI (see [#7465](https://github.com/danmar/cppcheck/pull/7465)) but it fails to run:

```
+ python3 iwyu-mapgen-apple-libc.py /Applications/Xcode_15.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
  File "/Users/runner/work/cppcheck/cppcheck/iwyu-mapgen-apple-libc.py", line 106
    print(f'# Generated using {' '.join(sys.argv)}')
                                                   ^
SyntaxError: f-string: expecting '}'
```
https://github.com/danmar/cppcheck/actions/runs/14433185399/job/40470455870

mohd-akram commented:
I think this is because of the nested quotes in the f-string. It requires Python 3.12+. Can be fixed by just using different quotes internally.

firewave commented:
That seems a bit high reaching given 3.13 was just released. Python 3.9 is still in support until October 2025: https://devguide.python.org/versions/.

I am using the GitHub `macos-13` runner which is still supported. According to the documentation that should be using 3.13 though: https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md ... and I confirmed that is indeed the case:

```
-- Python_VERSION =        3.13.2
```

firewave commented:
On a side note - maybe there should be build steps for testing the generators with the various available runners.

mohd-akram commented:
> That seems a bit high reaching given 3.13 was just released. Python 3.9 is still in support until October 2025: https://devguide.python.org/versions/.

Yeah, it should probably be changed. I didn't notice it actually.

> I am using the GitHub `macos-13` runner which is still supported. According to the documentation that should be using 3.13 though: https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md ... and I confirmed that is indeed the case:
> 
> ```
> -- Python_VERSION =        3.13.2
> ```

It should definitely work with that. I tested with Python 3.13.3.

firewave commented:
After removing that override the mapping generator is working.

But the analysis fails with an assertion:
```
/private/tmp/include-what-you-use-20250406-7581-b0uutx/include-what-you-use/iwyu_include_picker.cc:1467: Assertion failed: (*map)[key] == visibility Same file seen with two different visibilities: <sys/time.h> Old vis: 1 New vis: 2
```
https://github.com/danmar/cppcheck/actions/runs/14442622565/job/40495968319

[macos.imp.zip](https://github.com/user-attachments/files/19733484/macos.imp.zip)

----

A nit: the first line is a comment and IIRC mappings are JSON files. And JSON files do not allow comments. Some parsers actually do but the file is not valid.

kimgr commented:
@firewave The .imp format is not json, it's yaml, see the last paragraph of https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUMappings.md#mapping-files

kimgr commented:
@mohd-akram Maybe we could demodernize the python a bit for wider compat. How new are the Path API features in use? Those seem nice and useful, but I could live without the `f` formatting.

firewave commented:
> The .imp format is not json, it's yaml

Good to know before I had back into the Qt mapping generator.

FYI the LLVM YAML parser is not complete and ignored things silently in the past (like multi-line fields).

firewave commented:
> Maybe we could demodernize the python a bit for wider compat. How new are the Path API features in use? Those seem nice and useful, but I could live without the `f` formatting.

Both are already available in Python 3.6. The issue is some extension of the fstring syntax which was apparently only introduced in 3.12.

mohd-akram commented:
Thanks @firewave, good catch. I've simplified the script quite a bit in #1714 and moved more of the mappings to #1706.

@kimgr It should be widely compatible now in my PR. The earliest non-EOL Python is 3.9, that's also what comes with macOS and I've verified that the script runs with it well with the new changes.

firewave commented:
I did a comparison run and it looks good. Thanks!

kimgr commented on mapgen/iwyu-mapgen-apple-libc.py:
```diff
@@ -0,0 +1,129 @@
+
+        # map other headers
+        for glob, repl in PATTERN_MAPPINGS.items():
+            if p.relative_to(root).full_match(glob):
```
I noticed when testing the revised version that `Path.full_match` is not available until Python 3.13, so this fails to run on my fairly recent Ubuntu 24.04 system. Let's see if we can relax that requirement a bit when fixing the quotes issue too?

mohd-akram replied:
It can be changed to `Path.match` which is less strict but maybe is good enough.
